Do not init standard properties twice

rPraml opened the pull request:
I found a strange behaviour while fixing `instanceof` in #830.

My problem was, while `[] instanceof Array` retured true, `[] instanceof Object` returned false (at least in the NativeJavaListTest)

After some time of debugging, I noticed, that the NativeJavaListTest does the following:

```java
Global global = new Global();
global.init(ContextFactory.getGlobal());
Scriptable scope = context.initStandardObjects(global);
context.evaluateString(scope, "[] instanceof Object",...)
```
The problem is, that `global.init(ContextFactory.getGlobal());` creates all standard objects the first time, while `context.initStandardObjects(global)` will overwrite the standard objects. (Note: initStandardObjects will return the `global` value and no new instance)
So, we have 2 "Object" prototypes:
`scope.get("Object", scope).prototypeProperty` will return an other instance than `ScriptableObject.getObjectPrototype(scope)` 
This has to do something with 
```java
        if (scope instanceof TopLevel) {
            ((TopLevel) scope).cacheBuiltins(scope, sealed);
        }
```
at the end of the `initSafeStandardObjects` method.

In the first call of "initStandardObjects",, the cache is empty, so
- Object-Prototype is created (instance 1)
- Array-prototype is created (requires Object-prototype and gets instance 1 - no cache hit)
- ...
- cacheBuiltins is called.

on the second call of "initStandardObjects", 
- Object-Prototype is created (instance 2)
- Array-prototype is created (requires Object-prototype and gets instance 1 - from cache !)
- ...
- cacheBuiltins is called. (will update cached value to instance 2, but this is too late)


What would you suggest as solution:

**Solution 1:** Not calling "initStandardObjects" twice?
I would add an assert/exception in ScriptRuntime.initSafeStandardObjects:
```java
    public static ScriptableObject initSafeStandardObjects(
            Context cx, ScriptableObject scope, boolean sealed) {
        if (scope == null) {
            scope = new NativeObject();
        } else {
            assert !scope.has("Object",scope) : "already initialized";
        }
```

**Solution 2:** Clearing the cache, when "initStandardObjects" is called a second time:
```java
    public static ScriptableObject initSafeStandardObjects(
            Context cx, ScriptableObject scope, boolean sealed) {
        if (scope == null) {
            scope = new NativeObject();
        } else if (scope instanceof TopLevel) {
            ((TopLevel)scope).clearCache(); // will set TopLevel.ctors and TopLevel.errors to null
        }
```

in my oppinion, it does not make sense, to initialize objects twice, so throwing an exception/printing a warning would make sense, but may break existing code

rPraml commented on testsrc/org/mozilla/javascript/tests/JavaIterableTest.java:
```diff
@@ -146,19 +139,23 @@ public void testSet() {
+                .call(
+                        context -> {
+                            context.setLanguageVersion(Context.VERSION_ES6);
+                            Scriptable scope = context.newObject(global);
```
was `Scriptable scope = context.initStandardObjects(global);`
this is the only code change in this class, which is wrong, as it does not create a new scope, but initializes the standardObjects a second time.
I fixed this also in other tests

tonygermano replied:
I haven't tested it, but can that line be removed completely? I don't think it's necessary to create a new object as global already refers to an initialized scope. On the other hand, should a new global scope be initialized on each test instead of reusing the same one? I do see the problem you were trying to fix with the single global instance getting reinitialized on each test.

rPraml replied:
> I haven't tested it, but can that line be removed completely?

for this test case, it would work. Or in other words, it was just like that:
`context.initStandardObjects` returns `global` so the code looks like this:
```java
context.initStandardObjects(global);
Scriptable scope = global;
```

> On the other hand, should a new global scope be initialized on each test instead of reusing the same one?

it depends on the test. If the test itself modifies the scope, it would be better to initialize a new one. In this test, the next line puts `value` into the scope (`scope.put("value", scope, Context.javaToJS(value, scope));`) - as this is done on each test, I would agree, that it is not neccessary in this test, but the origin problem was, that `initStandardObjects` mess up some objects if it is called multiple times

rPraml commented:
@tonygermano I would suggest this [fix](https://github.com/mozilla/rhino/pull/1025/files#diff-58ee1701deb704ddb7d0ff0d41aef3275244d978ad279b1e4f1b4a61ff9e563d):

I've added a simpler test, that shows the originial problem more clearly: [InitStandardObjectTest](https://github.com/mozilla/rhino/pull/1025/files#diff-ccbc702c46f886e1b48e501b8d585cb480a2cac113435d0638a9ec5799dbf317R42)

The complex test [NestedContextPrototypeTest](https://github.com/mozilla/rhino/pull/1025/files#diff-ccbc702c46f886e1b48e501b8d585cb480a2cac113435d0638a9ec5799dbf317R42) showed me, that I am using `initStandardObjects` the wrong way (I can remove that test, if you propose, or keep it for documentation)

I also modified the NativeXXXTests, which have used `initStandardObjects` the wrong way. I think the intention of the original test writer @syjer was, to create a new isolated scope on each test run. So I would suggest, to keep this change. (Of course, the test will work without this line and act on `global` scope)

gbrail commented:
Thanks -- it looks like you went with the second option, which is to clear the cache so that calling "initStandardObjects" twice does not result in non-functional scripts.

I think that's a good choice, given that there is lots of code out there.

Also, thanks for the comprehensive tests!

If no one else has any objections then I'm OK with merging this.

rbri commented:
@gbrail lets merge....

gbrail commented:
Agreed -- this looks good. Thanks!
